pgrx/src/palloc/pbox.rs

```diff
@@ -0,0 +1,78 @@
+use crate::callconv::{BoxRet, FcInfo};
+use crate::datum::{BorrowDatum, Datum};
+use crate::layout::PassBy;
+use crate::memcx::MemCx;
+use crate::pg_sys;
+use core::marker::PhantomData;
+use core::ptr::NonNull;
+
+use pgrx_sql_entity_graph::metadata::{
+    ArgumentError, Returns, ReturnsError, SqlMapping, SqlTranslatable,
+};
+
+/** As [`Box<T, A>`][stdbox] where `A` is a [`MemCx`]
+
+
+[stdbox]: alloc::boxed::Box
+*/
+pub struct PBox<'mcx, T: ?Sized> {
+    ptr: NonNull<T>,
+    _cx: PhantomData<MemCx<'mcx>>,
+}
+
+impl<'mcx, T: ?Sized> PBox<'mcx, T> {
+    pub unsafe fn from_raw_in(ptr: NonNull<T>, _cx: &MemCx<'mcx>) -> PBox<'mcx, T> {
+        PBox { ptr, _cx: PhantomData }
+    }
+}
+
+impl<'mcx, T: Sized> PBox<'mcx, T> {
+    pub fn new_in(val: T, memcx: &MemCx<'mcx>) -> Self {
+        const { assert!(align_of::<T>() <= 8) };
```

Comment:
This can be relaxed on higher Postgres versions where we can use aligned allocations.

Reply:
instead of `<= 8` wouldn't we want `<= size_of::<usize>()`?

Reply:
Oh, true.

Reply:
I think this is correctest:
```suggestion
        const { assert!(align_of::<T>() <= size_of::<pg_sys::Datum>()) };
```